File: apps/macos/VoxtralRealtimeApp/VoxtralRealtime/Models/Session.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 * All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

struct Session: Identifiable, Codable, Sendable, Hashable {
    let id: UUID
    let date: Date
    var title: String
    var transcript: String
    var duration: TimeInterval

    init(
        id: UUID = UUID(),
        date: Date = .now,
        title: String = "",
        transcript: String = "",
        duration: TimeInterval = 0
    ) {
        self.id = id
        self.date = date
        self.title = title
        self.transcript = transcript
        self.duration = duration
    }

    var displayTitle: String {
        if !title.isEmpty { return title }
        let formatter = DateFormatter()
        formatter.dateStyle = .medium
        formatter.timeStyle = .short
        return formatter.string(from: date)
    }
}

File: parakeet/windows-cuda/ParakeetApp/ParakeetApp/Models/Session.cs
// Copyright (c) Meta Platforms, Inc. and affiliates.
// All rights reserved.
// This source code is licensed under the BSD-style license found in the
// LICENSE file in the root directory of this source tree.

using System.Text.Json.Serialization;

namespace ParakeetApp.Models;

public class TimestampSegment
{
    [JsonPropertyName("start")]
    public double Start { get; set; }

    [JsonPropertyName("end")]
    public double End { get; set; }

    [JsonPropertyName("text")]
    public string Text { get; set; } = "";
}

public class Session
{
    [JsonPropertyName("id")]
    public Guid Id { get; set; } = Guid.NewGuid();

    [JsonPropertyName("date")]
    public DateTime Date { get; set; } = DateTime.Now;

    [JsonPropertyName("title")]
    public string? Title { get; set; }

    [JsonPropertyName("transcript")]
    public string Transcript { get; set; } = "";

    [JsonPropertyName("duration")]
    public double Duration { get; set; }

    [JsonPropertyName("audioDuration")]
    public double AudioDuration { get; set; }

    [JsonPropertyName("timestamps")]
    public List<TimestampSegment> Timestamps { get; set; } = new();

    [JsonPropertyName("pinned")]
    public bool Pinned { get; set; }

    [JsonPropertyName("wavPath")]
    public string? WavPath { get; set; }

    [JsonIgnore]
    public string DisplayTitle => !string.IsNullOrEmpty(Title)
        ? Title
        : Date.ToString("MMM d, yyyy h:mm tt");

    [JsonIgnore]
    public string PreviewText
    {
        get
        {
            var text = Transcript.Trim();
            return text.Length > 100 ? text[..100] + "..." : text;
        }
    }

    [JsonIgnore]
    public string FormattedDuration
    {
        get
        {
            var ts = TimeSpan.FromSeconds(AudioDuration);
            return ts.TotalMinutes >= 1
                ? $"{(int)ts.TotalMinutes}:{ts.Seconds:D2}"
                : $"{ts.Seconds}s";
        }
    }
}
